### PR TITLE
*: support auto_increment_increment & auto_increment_offset.

### LINKED ISSUE #14245: autoid: support system variables `auto_increment_increment` and `auto_increment_offset`
## Feature Request

**Is your feature request related to a problem? Please describe:**
The system variable `auto_increment_increment` and `auto_increment_offset` are useful in some scenarios. For example, when the users want to do a Master-Master(bidirectional) replication with `auto_increment` columns, duplicate-key conflict and be avoided by setting different variable values in the clusters, for example:

Cluster 1:
```
mysql> create table t(a int key auto_increment);
mysql> set @@auto_increment_offset = 1
mysql> set @@auto_increment_increment = 2
mysql> insert into t values (...)
# the sequence of a will be 1, 3, 5, 7...
``` 
Cluster 2:
```
mysql> create table t(a int key auto_increment);
mysql> set @@auto_increment_offset = 2
mysql> set @@auto_increment_increment = 2
mysql> insert into t values (...)
# the sequence of a will be 2, 4, 6, 8...
```

For now, TiDB defines those 2 variables: users can set them but they don't actually take effect. We should make them work for those users who need them.

**Describe alternatives you've considered:**
The main implementation in MySQL Server can be found here:

https://github.com/mysql/mysql-server/blob/91a17cedb1ee880fe7915fb14cfd74c04e8d6588/sql/handler.cc#L3431

Basically, it made a rounding and then calculate next ID from the following formula:

>  nr = nr * auto_increment_increment + auto_increment_offset

The result can be guaranteed to be a multiple of `auto_increment_increment` offset with `auto_increment_offset`.

We can port this formula to TiDB's ID allocator:

https://github.com/pingcap/tidb/blob/0d7edc7ef32c52dd734c21dfade5860c71aececd/meta/autoid/autoid.go#L312

Somethings we may need to consider:

- The formula applies to the cached values only. The allocation of the next `step` from KV is not affected.
- The allocation of `rowid` should not be affected.
- The rebase behavior from explicit value insertion or `ALTER TABLE ... AUTO_INCREMENT` should not break the constraint of these 2 variables: looks it should be fine, the next value is calculated instantly by the formula above.

Some manual tests have been done from a raw prototype: https://github.com/bb7133/tidb/commit/adb83e3455dc78f7c0cf4d4952acd3091ccd529e.

**Teachability, Documentation, Adoption, Migration Strategy:**

This is simply missing compatibility feature, seems we don't need to care them.

### PR DESCRIPTION
**Pull request by AilinKid**
### What problem does this PR solve? 
Fix #14245 
Support auto_increment_increment & auto_increment_offset.

### What is changed and how it works?
1. Add increment&offset variable modify support.
2. Take increment & offset into allocate consideration.
3. Check MySQL doc and valid the range of these two variables.
4. Attach more tests to increment&offset allocation logic.

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported function/method change
func CalcNeededIDs(base, n, increment, offset int64, isUnsigned bool) int64 
 - Has interface methods change
Alloc(tableID int64, n uint64, increment, offset int64) (int64, int64, error)

Related changes

 - Need to update the documentation

Release note

 - Support system variable `auto_increment_increment` & `auto_increment_offset`.

### CODE REVIEW

File: meta/autoid/autoid_test.go [hunk 1]
```diff
@@ -323,10 +378,36 @@ func (*testSuite) TestUnsignedAutoid(c *C) {
...
+	c.Assert(uint64(max), Equals, uint64(math.MaxUint64-98))
+
+	c.Assert(max-min, Equals, autoid.CalcNeededIDs(int64(uint64(offset)-1), 2, increment, offset, true))
+	firstID := seekToFirstAutoID(min, increment, offset)
+	c.Assert(uint64(firstID), Equals, uint64(math.MaxUint64-100))
+
+}
+
+func seekToFirstAutoID(base, increment, offset int64) int64 {
+	nr := (base + increment - offset) / increment
```
Comment (wjhuang2016) [inline meta/autoid/autoid_test.go]: This formula appear many times, please use a function to wrap it.

File: sessionctx/variable/session.go [hunk 2]
```diff
@@ -932,6 +939,27 @@ func (s *SessionVars) SetSystemVar(name string, val string) error {
...
+		// AutoIncrementIncrement is valid in [1, 65535].
+		temp := tidbOptPositiveInt32(val, DefAutoIncrementIncrement)
+		if temp <= 0 {
+			s.AutoIncrementIncrement = 1
+		} else if temp > math.MaxUint16 {
+			s.AutoIncrementIncrement = math.MaxUint16
+		} else {
+			s.AutoIncrementIncrement = temp
+		}
+
```
Comment (wjhuang2016) [inline sessionctx/variable/session.go]: Remove the empty line.

File: sessionctx/variable/session.go [hunk 3]
```diff
@@ -932,6 +939,27 @@ func (s *SessionVars) SetSystemVar(name string, val string) error {
...
+	case AutoIncrementOffset:
+		// AutoIncrementOffset is valid in [1, 65535].
+		temp := tidbOptPositiveInt32(val, DefAutoIncrementOffset)
+		if temp <= 0 {
+			s.AutoIncrementOffset = 1
+		} else if temp > math.MaxUint16 {
+			s.AutoIncrementOffset = math.MaxUint16
+		} else {
+			s.AutoIncrementOffset = temp
+		}
```
Comment (wjhuang2016) [inline sessionctx/variable/session.go]: Can we use a function to wrap them?

File: executor/insert_test.go [hunk 4]
```diff
@@ -830,3 +830,47 @@ func (s *testSuite3) TestDMLCast(c *C) {
...
+func (s *testSuite3) TestAutoIDIncrementAndOffset(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec(`use test`)
+	// Test handle is PK.
+	tk.MustExec(`create table io (a int key auto_increment)`)
+	tk.Se.GetSessionVars().AutoIncrementOffset = 10
+	tk.Se.GetSessionVars().AutoIncrementIncrement = 2
+	tk.MustExec(`insert into io values (),(),()`)
+	tk.MustQuery(`select * from io`).Check(testkit.Rows("10", "12", "14"))
+	tk.MustExec(`delete from io`)
```
Comment (tangenta) [inline executor/insert_test.go]: It seems this behavior is incompatible with MySQL 8.0:
```sql
mysql root@172.17.0.2:test> create table t (col int key auto_increment)
Query OK, 0 rows affected
Time: 0.040s
mysql root@172.17.0.2:test> SET @@auto_increment_increment=2
Query OK, 0 rows affected
Time: 0.001s
mysql root@172.17.0.2:test> SET @@auto_increment_offset=10
Query OK, 0 rows affected
Time: 0.001s
mysql root@172.17.0.2:test> insert into t values (), (), ()
Query OK, 3 rows affected
Time: 0.016s
mysql root@172.17.0.2:test> select * from t;
+-----+
| col |
+-----+
| 2   |
| 4   |
| 6   |
+-----+
3 rows in set
Time: 0.007s
mysql root@172.17.0.2:test> 
```

As the documentation says, 
> When the value of auto_increment_offset is greater than that of auto_increment_increment, the value of auto_increment_offset is ignored.

Comment (AilinKid) [reply executor/insert_test.go]: yes, actually is a bug itself,
https://yq.aliyun.com/articles/294121
https://jira.mariadb.org/browse/MDEV-6839
Aili and Mariadb both complain about this, Cong and I wanna fix it.

File: meta/autoid/autoid.go [hunk 5]
```diff
@@ -319,13 +335,44 @@ func (alloc *allocator) Alloc(tableID int64, n uint64) (int64, int64, error) {
...
 	defer alloc.mu.Unlock()
 	if alloc.isUnsigned {
-		return alloc.alloc4Unsigned(tableID, n)
+		return alloc.alloc4Unsigned(tableID, n, increment, offset)
 	}
-	return alloc.alloc4Signed(tableID, n)
+	return alloc.alloc4Signed(tableID, n, increment, offset)
+}
+
+// CalcNeededIDs is exported for test.
```
[the remark this answers is not in the document]
Comment (AilinKid) [reply meta/autoid/autoid.go]: make sense.

File: executor/insert_common.go [hunk 6]
```diff
@@ -717,22 +717,24 @@ func (e *InsertValues) lazyAdjustAutoIncrementDatum(ctx context.Context, rows []
...
-				e.lastInsertID = uint64(min) + 1
+				e.lastInsertID = uint64(min)
 			}
 			// Assign autoIDs to rows.
 			for j := 0; j < cnt; j++ {
 				offset := j + start
 				d := rows[offset][colIdx]
 
-				id := int64(uint64(min) + uint64(j) + 1)
+				id := min + int64(j)*increment
```
[the remark this answers is not in the document]
Comment (AilinKid) [reply executor/insert_common.go]: good catch

File: executor/insert_test.go (hunk 4 above)
Comment (bb7133) [reply executor/insert_test.go]: Good catch @tangenta 

The behavior of MySQL in such cases is actually random, that is terrible. Fixing it is a better idea, and we should point this out in the official document later. @AilinKid

File: meta/autoid/autoid.go [hunk 7]
```diff
@@ -319,13 +335,44 @@ func (alloc *allocator) Alloc(tableID int64, n uint64) (int64, int64, error) {
...
-	n1 := int64(n)
+func (alloc *allocator) alloc4Signed(tableID int64, n uint64, increment, offset int64) (int64, int64, error) {
+	// Check offset rebase if necessary.
+	if offset-1 > alloc.base {
+		if err := alloc.rebase4Signed(tableID, offset-1, true); err != nil {
+			return 0, 0, err
+		}
+	}
+	// Calculate the total batch size needed.
+	n1 := CalcNeededIDs(alloc.base, int64(n), increment, offset, alloc.isUnsigned)
```
[the remark this answers is not in the document]
Comment (AilinKid) [reply meta/autoid/autoid.go]: Addressed

File: meta/autoid/autoid_test.go (hunk 1 above)
Comment (AilinKid) [reply meta/autoid/autoid_test.go]: Addressed

File: executor/insert_test.go [hunk 8]
```diff
@@ -830,3 +830,47 @@ func (s *testSuite3) TestDMLCast(c *C) {
 	tk.MustExec("delete from t where a = ''")
 	tk.MustQuery(`select * from t`).Check(testkit.Rows())
 }
+
+func (s *testSuite3) TestAutoIDIncrementAndOffset(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
```
[the remark this answers is not in the document]
Comment (bb7133) [reply executor/insert_test.go]: Please test the valid range for the 2 variables:

> Both variables have global and session values, and each can assume an integer value between 1 and 65,535 inclusive. Setting the value of either of these two variables to 0 causes its value to be set to 1 instead.

File: meta/autoid/autoid.go [hunk 9]
```diff
@@ -80,7 +80,11 @@ type Allocator interface {
 	// Alloc allocs N consecutive autoID for table with tableID, returning (min, max] of the allocated autoID batch.
 	// It gets a batch of autoIDs at a time. So it does not need to access storage for each call.
 	// The consecutive feature is used to insert multiple rows in a statement.
-	Alloc(tableID int64, n uint64) (int64, int64, error)
+
+	// increment & offset is used to valid the start position (the allocator's base is not always the last allocated id).
+	// The returned range is still (min, max], you need derive the ids like firstID, firstID + increment * 2... in caller.
```
Comment (bb7133) [inline meta/autoid/autoid.go]:
```suggestion
	// The returned range is still (min, max], you need derive the ids like firstID, firstID + increment * 2... in caller.
```
Please update the comment, "the returned range is _still (min,max]_" is not proper now.

File: meta/autoid/autoid.go [hunk 10]
```diff
@@ -309,7 +313,19 @@ func NewAllocatorsFromTblInfo(store kv.Storage, schemaID int64, tblInfo *model.T
 }
 
 // Alloc implements autoid.Allocator Alloc interface.
-func (alloc *allocator) Alloc(tableID int64, n uint64) (int64, int64, error) {
+// For autoIncrement allocator, the increment and offset should always be positive in [1, 65535].
+// Attention:
+// When increment and offset is not the default value = 1, the return range (min, max] need to
```
Comment (bb7133) [inline meta/autoid/autoid.go]: Ditto, you may need to update (min, max]

File: meta/autoid/autoid.go [hunk 11]
```diff
@@ -319,13 +335,44 @@ func (alloc *allocator) Alloc(tableID int64, n uint64) (int64, int64, error) {
...
 	if alloc.isUnsigned {
-		return alloc.alloc4Unsigned(tableID, n)
+		return alloc.alloc4Unsigned(tableID, n, increment, offset)
 	}
-	return alloc.alloc4Signed(tableID, n)
+	return alloc.alloc4Signed(tableID, n, increment, offset)
+}
+
+// CalcNeededIDs is exported for test.
+func CalcNeededIDs(base, n, increment, offset int64, isUnsigned bool) int64 {
```
Comment (bb7133) [inline meta/autoid/autoid.go]: I think we should rename this function, `CalcNeededIDs` is not a good name.

File: meta/autoid/autoid.go (hunk 9 above)
Comment (AilinKid) [reply meta/autoid/autoid.go]: the allocate logic is still the same, return (min, max],
the caller's responsibility is to derive the autoid series from this range
 refer: `AllocBatchAutoIncrementValue`

File: executor/insert_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (AilinKid) [reply executor/insert_test.go]: addressed.

File: meta/autoid/autoid.go (hunk 11 above)
Comment (AilinKid) [reply meta/autoid/autoid.go]: Addressed.

File: executor/insert_common.go [hunk 12]
```diff
@@ -717,22 +717,24 @@ func (e *InsertValues) lazyAdjustAutoIncrementDatum(ctx context.Context, rows []
 				cnt++
 			}
 			// Alloc batch N consecutive (min, max] autoIDs.
```
[the remark this answers is not in the document]
Comment (AilinKid) [reply executor/insert_common.go]: still return (min, max], if n = 1, the allocated id is max, if n != 1, the caller should derive auto series from the range.

File: sessionctx/variable/session.go (hunk 2 above)
Comment (AilinKid) [reply sessionctx/variable/session.go]: addressed.

File: sessionctx/variable/session.go (hunk 3 above)
Comment (AilinKid) [reply sessionctx/variable/session.go]: it's ok here.

File: executor/insert_test.go [hunk 13]
```diff
@@ -830,3 +830,71 @@ func (s *testSuite3) TestDMLCast(c *C) {
 	tk.MustExec("delete from t where a = ''")
 	tk.MustQuery(`select * from t`).Check(testkit.Rows())
 }
+
+// TiDB has fixed a issue, when the value of auto_increment_offset is greater than that of auto_increment_increment,
```
Comment (bb7133) [inline executor/insert_test.go]: How about changing it to:

```suggestion
// TiDB has fixed a issue, when the value of auto_increment_offset is greater than that of auto_increment_increment,
```
"There is a potential issue in MySQL: when the value of auto_increment_offset is greater than that of auto_increment_increment, the value of auto_increment_offset is ignored(https://dev.mysql.com/doc/refman/8.0/en/replication-options-master.html#sysvar_auto_increment_increment), 
This issue is a flaw of the implementation of MySQL and it doesn't exist in TiDB"

File: executor/insert_common.go [hunk 14]
```diff
@@ -716,23 +716,23 @@ func (e *InsertValues) lazyAdjustAutoIncrementDatum(ctx context.Context, rows []
 				i++
 				cnt++
 			}
-			// Alloc batch N consecutive (min, max] autoIDs.
-			// max value can be derived from adding one for cnt times.
-			min, _, err := table.AllocBatchAutoIncrementValue(ctx, e.Table, e.ctx, cnt)
+			// Alloc batch N consecutive autoIDs, returning min and increment.
```
Comment (wjhuang2016) [inline executor/insert_common.go]: `returning min and increment.` I think this comment is redundant, the code below this already said so.

File: sessionctx/variable/session.go (hunk 3 above)
Comment (wjhuang2016) [reply sessionctx/variable/session.go]: Line 944-951 and Line 953 to 961 are equal, we shouldn't let the code duplicate

File: meta/autoid/autoid_test.go (hunk 1 above)
Comment (wjhuang2016) [reply meta/autoid/autoid_test.go]: I found used `nr := ...`  four times , please update other usages

File: meta/autoid/errors.go [hunk 15]
```diff
@@ -20,11 +20,12 @@ import (
...
-	ErrAutoRandReadFailed   = terror.ClassAutoid.New(mysql.ErrAutoRandReadFailed, mysql.MySQLErrName[mysql.ErrAutoRandReadFailed])
+	errInvalidTableID            = terror.ClassAutoid.New(mysql.ErrInvalidTableID, mysql.MySQLErrName[mysql.ErrInvalidTableID])
+	errInvalidIncrementAndOffset = terror.ClassAutoid.New(mysql.ErrInvalidIncrementAndOffset, mysql.MySQLErrName[mysql.ErrInvalidIncrementAndOffset])
+	ErrAutoincReadFailed         = terror.ClassAutoid.New(mysql.ErrAutoincReadFailed, mysql.MySQLErrName[mysql.ErrAutoincReadFailed])
+	ErrWrongAutoKey              = terror.ClassAutoid.New(mysql.ErrWrongAutoKey, mysql.MySQLErrName[mysql.ErrWrongAutoKey])
+	errInvalidAllocatorType      = terror.ClassAutoid.New(mysql.ErrUnknownAllocatorType, mysql.MySQLErrName[mysql.ErrUnknownAllocatorType])
+	ErrAutoRandReadFailed        = terror.ClassAutoid.New(mysql.ErrAutoRandReadFailed, mysql.MySQLErrName[mysql.ErrAutoRandReadFailed])
 )
 
 func init() {
```
[the remark this answers is not in the document]
Comment (AilinKid) [reply meta/autoid/errors.go]: Good catch!

File: executor/insert_test.go (hunk 13 above)
Comment (AilinKid) [reply executor/insert_test.go]: nice, idea!

File: executor/insert_common.go (hunk 14 above)
Comment (AilinKid) [reply executor/insert_common.go]: nice, addressed.

File: meta/autoid/autoid.go [hunk 16]
```diff
@@ -309,23 +313,77 @@ func NewAllocatorsFromTblInfo(store kv.Storage, schemaID int64, tblInfo *model.T
 }
 
 // Alloc implements autoid.Allocator Alloc interface.
-func (alloc *allocator) Alloc(tableID int64, n uint64) (int64, int64, error) {
+// For autoIncrement allocator, the increment and offset should always be positive in [1, 65535].
+// Attention:
+// When increment and offset is not the default value(1), the return range (min, max] need to
+// calculate the right start position rather than simply the add 1 to min. Then you can derive
```
Comment (wjhuang2016) [inline meta/autoid/autoid.go]: remove the second `the`

File: meta/autoid/autoid.go [hunk 17]
```diff
@@ -378,14 +436,22 @@ func (alloc *allocator) alloc4Signed(tableID int64, n uint64) (int64, int64, err
...
-	n1 := int64(n)
+func (alloc *allocator) alloc4Unsigned(tableID int64, n uint64, increment, offset int64) (int64, int64, error) {
+	// Check offset rebase if necessary.
+	if uint64(offset-1) > uint64(alloc.base) {
+		if err := alloc.rebase4Unsigned(tableID, uint64(offset-1), true); err != nil {
+			return 0, 0, err
+		}
+	}
+	// Calculate the total batch size needed.
+	n1 := CalcNeededBatchSize(alloc.base, int64(n), increment, offset, alloc.isUnsigned)
```
Comment (wjhuang2016) [inline meta/autoid/autoid.go]: Please rename `n1`

File: sessionctx/variable/session.go [hunk 18]
```diff
@@ -384,6 +385,10 @@ type SessionVars struct {
 
 	SQLMode mysql.SQLMode
 
+	AutoIncrementIncrement int
```
Comment (wjhuang2016) [inline sessionctx/variable/session.go]: Please add comment.

File: meta/autoid/autoid.go (hunk 17 above)
Comment (AilinKid) [reply meta/autoid/autoid.go]: it's ok here, same to previous code.

File: meta/autoid/autoid.go [hunk 19]
```diff
@@ -80,7 +80,11 @@ type Allocator interface {
 	// Alloc allocs N consecutive autoID for table with tableID, returning (min, max] of the allocated autoID batch.
 	// It gets a batch of autoIDs at a time. So it does not need to access storage for each call.
 	// The consecutive feature is used to insert multiple rows in a statement.
-	Alloc(tableID int64, n uint64) (int64, int64, error)
+
+	// increment & offset is used to valid the start position (the allocator's base is not always the last allocated id).
```
Comment (zimulala) [inline meta/autoid/autoid.go]: The function comment begins with the function name. This funcion comment like `Alloc ...`.

Comment (AilinKid) [reply meta/autoid/autoid.go]: addressed.

File: meta/autoid/autoid.go [hunk 20]
```diff
@@ -309,23 +312,86 @@ func NewAllocatorsFromTblInfo(store kv.Storage, schemaID int64, tblInfo *model.T
...
 	defer alloc.mu.Unlock()
 	if alloc.isUnsigned {
-		return alloc.alloc4Unsigned(tableID, n)
+		return alloc.alloc4Unsigned(tableID, n, increment, offset)
+	}
+	return alloc.alloc4Signed(tableID, n, increment, offset)
+}
+
+func validIncrementAndOffset(increment, offset int64) bool {
+	return (increment >= 1 && increment <= 65535) && (offset >= 1 && offset <= 65535)
```
Comment (wjhuang2016) [inline meta/autoid/autoid.go]: Please use a constant instead of magic number.

File: table/table.go [hunk 21]
```diff
@@ -217,20 +217,32 @@ func AllocAutoIncrementValue(ctx context.Context, t Table, sctx sessionctx.Conte
...
 	}
-	return t.Allocator(sctx, autoid.RowIDAllocType).Alloc(t.Meta().ID, uint64(N))
+	increment = int64(sctx.GetSessionVars().AutoIncrementIncrement)
+	offset := int64(sctx.GetSessionVars().AutoIncrementOffset)
+	min, max, err := t.Allocator(sctx, autoid.RowIDAllocType).Alloc(t.Meta().ID, uint64(N), int64(increment), int64(offset))
+	if err != nil {
+		return min, max, err
+	}
+	// seek to first autoID.
+	nr := (min + increment - offset) / increment
```
Comment (wjhuang2016) [inline table/table.go]: Use the function

File: table/table.go [hunk 22]
```diff
@@ -217,20 +217,32 @@ func AllocAutoIncrementValue(ctx context.Context, t Table, sctx sessionctx.Conte
...
+// The caller can derive the autoID by adding increment to firstID for N-1 times.
+func AllocBatchAutoIncrementValue(ctx context.Context, t Table, sctx sessionctx.Context, N int) (firstID int64, increment int64, err error) {
 	if span := opentracing.SpanFromContext(ctx); span != nil && span.Tracer() != nil {
 		span1 := span.Tracer().StartSpan("table.AllocBatchAutoIncrementValue", opentracing.ChildOf(span.Context()))
 		defer span1.Finish()
 	}
-	return t.Allocator(sctx, autoid.RowIDAllocType).Alloc(t.Meta().ID, uint64(N))
+	increment = int64(sctx.GetSessionVars().AutoIncrementIncrement)
+	offset := int64(sctx.GetSessionVars().AutoIncrementOffset)
+	min, max, err := t.Allocator(sctx, autoid.RowIDAllocType).Alloc(t.Meta().ID, uint64(N), int64(increment), int64(offset))
```
Comment (wjhuang2016) [inline table/table.go]: int64()  is redundant.

File: executor/insert_common.go [hunk 23]
```diff
@@ -716,23 +716,23 @@ func (e *InsertValues) lazyAdjustAutoIncrementDatum(ctx context.Context, rows []
...
+			// max value can be derived from adding the increment value to min for cnt-1 times.
+			min, increment, err := table.AllocBatchAutoIncrementValue(ctx, e.Table, e.ctx, cnt)
 			if e.handleErr(col, &autoDatum, cnt, err) != nil {
 				return nil, err
 			}
 			// It's compatible with mysql setting the first allocated autoID to lastInsertID.
 			// Cause autoID may be specified by user, judge only the first row is not suitable.
 			if e.lastInsertID == 0 {
-				e.lastInsertID = uint64(min) + 1
+				e.lastInsertID = uint64(min)
```
[the remark this answers is not in the document]
Comment (AilinKid) [reply executor/insert_common.go]: we already derive the firstID and increment from the allocated range.
`AllocBatchAutoIncrementValue` now return firstID and increment.

File: table/table.go (hunk 21 above)
Comment (AilinKid) [reply table/table.go]: addressed

File: meta/autoid/autoid.go (hunk 20 above)
Comment (AilinKid) [reply meta/autoid/autoid.go]: addressed

File: sessionctx/variable/session.go (hunk 3 above)
Comment (AilinKid) [reply sessionctx/variable/session.go]: addressed

File: meta/autoid/autoid.go [hunk 24]
```diff
@@ -75,12 +77,15 @@ const (
 )
 
 // Allocator is an auto increment id generator.
-// Just keep id unique actually.
+// Just keep id unique cvtually.
```
[the remark this answers is not in the document]
Comment (AilinKid) [reply meta/autoid/autoid.go]: addressed

**Comment by bb7133** [comment]
LGTM

**Pull request merged**